File: Python3/70. Climbing Stairs.py
# Recursive Solution
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1:
            return 1
        
        elif n == 2:
            return 2
        
        # ways[n] = ways[n-1] + ways[n-2]
        ways = [-1]*(n+1)
        ways[1] = 1
        ways[2] = 2
        
        def rec(n):
            if ways[n] != -1:
                return ways[n]
            
            ways[n-1] = rec(n-1)
            ways[n-2] = rec(n-2)
            
            return ways[n-1] + ways[n-2]
            
        return rec(n)

File: Java/70. Climbing Stairs.java
class Solution {
    public int climbStairs(int n) {
        // ways[n] = ways[n-1] + ways[n-2]
        int first = 1;
        int second = 2;
        
        for(int i=1; i<n-1; i++){
            int temp = first+second;
            first = second;
            second = temp;
        }
        
        return n == 1 ? first : second;
    }
}